### ISSUE TITLE
System.Data.Odbc ERROR [IM002] [Microsoft] [ODBC driver manager]

### ISSUE THREAD

**Issue by 2881099** [dotnet/runtime#772]
Error [im002] [Microsoft] [ODBC driver manager] No data source name found and no default driver specified

![image](https://user-images.githubusercontent.com/16286519/70633983-65e87400-1c6c-11ea-8d30-eabc2949cb4f.png)

Environment: netcoreapp 3.1

Please see the nuget package upgrade below. There will be no errors before the upgrade. The above errors occur after the upgrade.

![image](https://user-images.githubusercontent.com/16286519/70635485-f7f17c00-1c6e-11ea-9461-476f49ca73f5.png)

Error :
> Driver={MySQL ODBC 8.0 Unicode Driver};Server=127.0.0.1;...

Success:
> Driver={SQL Server};Server=.;...

Only SQL Server succeeded, all other errors.

**Comment by roji** [comment]
@2881099 this issue is lacking some important information. Can you please clarify exactly what you're trying to do, and preferably submit a minimal code sample which illustrates the problem?

**Comment by 2881099** [comment]
![image](https://user-images.githubusercontent.com/16286519/70766993-93334000-1d9a-11ea-931b-2962325807ed.png)

Test Oracle ODBC in the picture.

If SQL Server ODBC is tested successfully.
Only SQL Server succeeded, all other errors.

Exception StackTrace:

```text
   在 System.Data.Odbc.OdbcConnection.HandleError(OdbcHandle hrHandle, RetCode retcode)
   在 System.Data.Odbc.OdbcConnectionHandle..ctor(OdbcConnection connection, OdbcConnectionString constr, OdbcEnvironmentHandle environmentHandle)
   在 System.Data.Odbc.OdbcConnectionOpen..ctor(OdbcConnection outerConnection, OdbcConnectionString connectionOptions)
   在 System.Data.Odbc.OdbcConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningObject)
   在 System.Data.ProviderBase.DbConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)
   在 System.Data.ProviderBase.DbConnectionFactory.CreateNonPooledConnection(DbConnection owningConnection, DbConnectionPoolGroup poolGroup, DbConnectionOptions userOptions)
   在 System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   在 System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   在 System.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   在 System.Data.ProviderBase.DbConnectionInternal.OpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory)
   在 System.Data.Odbc.OdbcConnection.Open()
```

**Comment by 2881099** [comment]
[debug_odbc.zip](https://github.com/dotnet/runtime/files/3958905/debug_odbc.zip)

This is a well organized debugging project.

**Comment by 2881099** [comment]
```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.9.0" />
    <PackageReference Include="xunit" Version="2.4.0" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
  </ItemGroup>
```

Using these versions is normal. @roji

**Comment by 2881099** [comment]
The above error prompt occurred in the latest version of vs2019.
I try to execute the test through the "dotnet test" console, as if there is no error. @roji

**Comment by danmosemsft** [comment]
What happens if you try from a .NET Framework project?

**Comment by 2881099** [comment]
[debug_odbc.zip](https://github.com/dotnet/runtime/files/3958905/debug_odbc.zip)

@danmosemsft "netcoreapp 3.1" test project

**Comment by danmosemsft** [comment]
@2881099 right, I’m asking whether you can also try .NET Framework , same code and same machine and tell us whether that works

**Comment by 2881099** [comment]
![image](https://user-images.githubusercontent.com/16286519/70770837-f7a9cb80-1da9-11ea-9bdd-bdfad0154c1f.png)

@danmosemsft . net framework 4.6.1 same error

**Comment by roji** [comment]
> There will be no errors before the upgrade. The above errors occur after the upgrade.

@2881099 can you clarify exactly which package you upgraded? Which version worked, and which version caused the failure?

**Comment by roji** [comment]
>  I try to execute the test through the "dotnet test" console, as if there is no error. @roji

Also, can you please confirm this? Are you saying that when you run the test from the console (with dotnet test), it always passes, and fails only when the test is executed from the Visual Studio test runner?

**Comment by 2881099** [comment]
> Also, can you please confirm this? Are you saying that when you run the test from the console (with dotnet test), it always passes, and fails only when the test is executed from the Visual Studio test runner?

Yes. 

[Please look debug_odbc.zip](https://github.com/dotnet/runtime/files/3958905/debug_odbc.zip)
